RFC: Unsafe Set Enum Discriminants

jamesmunns opened the pull request:
# Summary

This RFC proposes a way to write the discriminant of an enum when building it "from scratch". This introduces two new library components, an unsafe `set_discriminant` function, and a `discriminant_of!` macro, which can be used with any enum, regardless of `repr` or any other details.

This RFC is a follow-on to the `offset_of_enum` feature, tracked by https://github.com/rust-lang/rust/issues/120141.

jamesmunns commented:
I've gotten feedback from @RalfJung in Zulip that it's probably more reasonable to have `set_discriminant()` take an `*mut T` rather than an `&mut MaybeUninit<T>`, and I think that is reasonable. I'll probably update this draft to switch these in a bit, if there are no counter-objections.

clarfonthey commented on text/3727-unsafe-set-discriminant.md:
How exactly would this work, anyway? 

The presumption here is that `Discriminant<T>` only directly covers the discriminants of the type `T`, and not its children, so, I'm not even sure what this kind of expression would return. Would it return `Discriminant<Option<u32>>`?

jamesmunns replied:
Yep, I don't want to get too much into this (I wanted to leave it out for simplicity), but I would expect that second one to return `Discriminant<Option<u32>>`.

clarfonthey replied:
I mostly point it out since it feels disingenuous to say it's simply "not allowed yet" and rather that it doesn't make as much sense as an operation, and would need a proper design first.

jamesmunns replied:
I think we're on the same page! I'm saying "the implementor of this macro doesn't need to worry about that possibility, as we haven't specified it outside of 'don't do that'".

If you can think of a better way to word that to impart the meaning of "not defined within the scope of this RFC", I'm open to change it!

clarfonthey commented on text/3727-unsafe-set-discriminant.md:
Personally, I would expect `mem::set_discriminant` to be paired with a `ptr::set_discriminant` that accepts a pointer instead of a reference, and so it would logically make sense that there also be a `ptr::set_discriminant_unaligned` as well.

RalfJung replied:
FWIW `set_discriminant_unaligned` is not something that currently exists in the compiler at all.

Note that `read/write_unaligned` are implemented entirely in the library, they don't need compiler support.

So supporting unaligned discriminant writes would be a much bigger change from the implementation side. I'd keep this as a future possibility for now. (And if we get  unaligned discriminant writes we surely should also get  unaligned discriminant reads.)

clarfonthey replied:
Ah, that's fair. I would imagine that it would be possible to do with an unaligned read followed by an unaligned write, but it would require a bit of extra support from the compiler to ensure that you only read/write the part that contains the discriminant, rather than the entire type (which could be quite large).

jamesmunns replied:
Noting that I don't have opinions about where the methods end up, I would defer to the desires of the lib/lang teams.

jamesmunns replied:
I'll also add a note to "future potential" about unaligned discriminant reads as well.

clarfonthey replied:
I mostly point out the option to have multiple methods because it feels like a very reasonable answer to "should this method accept a pointer or a reference" is "why not both?" since there's definitely precedent to do this. You're right that it's not super important right now, but it lends itself to the fact that the actual signature isn't nearly as important to the RFC as the actual functionality, since we can offer either or both signatures in the final version.

jamesmunns commented on text/3727-unsafe-set-discriminant.md:
That's a fair point. I could relax this MUST a bit if you are working with specifically known types, and state that (for example) with generic code, you must always set it, and that setting it explicitly is never harmful (and will be a no-op in cases like this so there is no penalty for doing it).

clarfonthey replied:
Yeah, I definitely agree with encouraging people to use it anyway, although my main point here is that this RFC shouldn't retroactively require it when people were soundly using it before: we have defined cases where it was correct to simply write fields, and this adds in extra cases where it was not possible to do correctly before.

RalfJung replied:
> that setting it explicitly is never harmful (and will be a no-op in cases like this so there is no penalty for doing it).

No it's not a NOP. It reads the discriminant and causes UB if that produces the wrong result.

A NOP can never be UB.

jamesmunns commented:
Addressed all review comments so far. Link to the diff of changes since the initial version: https://github.com/rust-lang/rfcs/pull/3727/files/5c74b691972d36d30e27f253e8e2673d9930316f..eb486ef529eea98b7d2d9b0ffcd2254ac16a4ea0

coolreader18 commented:
What's the current legality of initializing niche-optimized enums?
```rust
let mut x: Option<NonNull<u8>> = None;
(&raw mut x).cast::<NonNull<u8>>().write(pre); // UB
```
I suppose there are guarantees made about the layout of Option for FFI purposes, but this feels... sketchy. (and maybe miri already has an issue with this?) That to say, unless these semantics are already defined, I feel like the section about calling set_discriminant on niche-optimized enums should be strengthened from "encouraged for explicitness" to "should"/"must". Maybe with an exception for Option/Result-like enums with a single niche at 0, since FFI-compatibility guarantees are already made for those, but I'd err on not allowing it if possible.

matthieu-m commented:
I was thinking about the (possible) future alternative receivers for `set_discriminant`:

 - Having to call `.as_mut_ptr()` is not so bad, what is more annoying is the loss of guarantees. Yes my `&mut MaybeUninit` has given me a non-null & sufficiently-aligned pointer, thank you very much. This is the true boilerplate.
 - Having `set_discriminant_mu` and `set_discriminant_nn` (or whatever) feels... heavy. Remembering the specific names is going to be annoying.
 - A `SetDiscriminant::set_discriminant` trait or a `set_discriminant<T: SealedDiscriminant>(t: T, ..)` method muddle safety guarantees: it makes it harder, at the call site, to ensure that the expected receiver is used, and it makes it brittle with regard to refactorings. Change the "receiver" from `NonNull<T>` to `*mut T`? Make sure the pointer is non-null now!
 - Inherent methods on `NonNull` and `MaybeUninit` directly is another possibility. If just called `set_discriminant` they would have some of the disadvantages of the trait-based approach, but for `MaybeUninit` a `write_discriminant` inherent method could be more in keeping with the current `write...` API and the different name would prevent a maintenance mistake.

So, all in all, for now I see two alternatives I like:

 - Don't introduce any variant. Bit boilerplatey in safety guarantees, but very explicit, and explicit is not so bad for `unsafe`.
 - Only introduce a single variant: `MaybeUninit::<T>::write_discriminant(&mut self, disc: Discriminant<T>)`. It's natural to have, and offers a lot of guarantees that `*mut T` doesn't -- non null, sufficiently aligned, exclusive access -- leaving only "correctly initialized" to fulfill.

Imberflur commented on text/3727-unsafe-set-discriminant.md:
It looks like the guarantee was extended to "Option-like" enums [recently](https://github.com/rust-lang/rust/pull/130628#issuecomment-2402761599) and there are some in progress [documentation](https://github.com/rust-lang/reference/pull/1654) [updates](https://github.com/rust-lang/unsafe-code-guidelines/pull/538).

jamesmunns commented on text/3727-unsafe-set-discriminant.md:
In the original draft, I had proposed using `&mut MaybeUninit<T>`, rather than `*mut T`. When I swapped that over I left this section, but refactored it. I probably could remove it, but there is still some open discussion about what API should there be on day one.

I'll likely remove this in the near future, as I am somewhat convinced that if there were to be only one interface, it should be on `*mut T`.

RalfJung commented on text/3727-unsafe-set-discriminant.md:
Rust does have other types that have discriminants, namely the generators that are created by lowering `async`. So we have to be a bit careful in how this is worded.

RalfJung commented on text/3727-unsafe-set-discriminant.md:
This sounds about right, though I would use some of the terms a bit differently. Specifically, set_discriminant will never read or write anything but the discriminant. However, enum layout is unspecified, and the discriminant can overlap with arbitrary values of the body.

It may be worth spelling out exactly what the safety requirements are: when writing the discriminant of variant `V`, then *all fields of `V` that overlap with the discriminant must be set to values that satisfy the language invariant of their type*. In practice, we don't guarantee which fields overlap with the discriminant, so *all* fields must be properly initialized.

This can replace the discussion of "read back" etc, that is just an implementation detail. Likewise, you can keep the example setting `None`, but I would frame it differently: because `None` has no fields, the requirement is trivially met. So the instruction is not "responsible for also initializing the whole body of the variant", all it really has to ensure is that future reads of the discriminant behave correctly.

> It'd be good also for this (or the following) section to be extended with more details about why this is the correct semantic.

The reason we introduced the readback is to make it so that
```
SetDiscriminant(place, variant_idx);
_x = GetDiscriminant(place);
```
can be optimize to
```
SetDiscriminant(place, variant_idx);
_x = variant_idx;
```
This seems like an obviously desirable transformation, so we must ensure that e.g. writing the `Some` discriminant can only return successfully if the next `GetDiscriminant` indeed would return `Some`.

joshlf replied:
@RalfJung Would the optimization you have in mind be impeded if we merely say that any enum value which is not wrapped in `MaybeUninit` is valid? Ie, that initialization of the variant and fields may happen in any order, but that initialization must be complete by the time that `MaybeUninit::assume_init` is called? That would permit `set_discriminant` to not need to constrain the existing content of the enum since it would basically just be a fancy `ptr::write`, and it would make it easier to reason about for callers. I assume there's a reason that this simpler approach isn't viable, but I'm curious what that reason is.

RalfJung replied:
I don't understand the question. It seems you are proposing to just not do the optimization. But we think we should reserve the possibility to do that optimization, and then it follows immediately that we need the UB.

IMO it'd also be rather strange to say that setting the discriminant of `Option<&T>` to `Some` on an all-null value is just a NOP and permitted. The programmer asked us to do something nonsensical.

y86-dev commented:
I would really like to have this feature to be able to support enums in [`pin-init`](https://github.com/Rust-for-Linux/pin-init/issues/59). At Rust-for-Linux, we have a use-case that would benefit from being able to initialize enums in-place.

As for the API, just having the `unsafe fn set_discriminant<T: ?Sized>(*mut T, Discriminant<T>)` would be enough. `pin-init` wouldn't expose it to users, as it would use that under the hood in macros to initialize the correct variant.

jamesmunns commented:
Noting that we had a design meeting about this, and the outcome was that I'll be updating the RFC to:

* Refocus the macro to remove the proposed `get_discriminant!` macro and `set_discriminant` function, instead only proposing a single `set_discriminant!` macro:
  * Has the form `set_discriminant!(PLACE, VARIANT) -> &mut T`, where T is the T of the PLACE
* the current `get_discriminant!` macro and `set_discriminant` method, as well as potentially other inherent methods will be moved to "future work"

However, I think answering the questions around `offset_of_enum` take precedence to working on this RFC, as it is a "hard prerequisite" to enabling this RFC. See https://github.com/rust-lang/rust/pull/143954, and particularly answer the questions outlined in https://github.com/rust-lang/rust/pull/143954#issuecomment-3084929032.

I'll likely defer updating this RFC until the `offset_of_enum` stabilization process seems decided.
